packages/cli/src/push/index.ts

```diff
@@ -44,44 +41,96 @@ const MAX_PAYLOAD_SIZE_BYTES = 5 * 1024 * 1024; // 5 MiB
  */
 @Service()
 export class Push extends TypedEmitter<PushEvents> {
-	isBidirectional = useWebSockets;
+	private useWebSockets = this.config.backend === 'websocket';
+
+	isBidirectional = this.useWebSockets;
 
-	private backend = useWebSockets ? Container.get(WebSocketPush) : Container.get(SSEPush);
+	private backend = this.useWebSockets ? Container.get(WebSocketPush) : Container.get(SSEPush);
 
 	constructor(
+		private readonly config: PushConfig,
 		private readonly instanceSettings: InstanceSettings,
-		private readonly publisher: Publisher,
 		private readonly logger: Logger,
+		private readonly authService: AuthService,
+		private readonly publisher: Publisher,
 	) {
 		super();
 		this.logger = this.logger.scoped('push');
 
-		if (useWebSockets) this.backend.on('message', (msg) => this.emit('message', msg));
+		if (this.useWebSockets) this.backend.on('message', (msg) => this.emit('message', msg));
 	}
 
 	getBackend() {
 		return this.backend;
 	}
 
+	/** Sets up the main express app to upgrade websocket connections */
+	setupPushServer(restEndpoint: string, server: Server, app: Application) {
+		if (this.useWebSockets) {
+			const wsServer = new WSServer({ noServer: true });
+			server.on('upgrade', (request: WebSocketPushRequest, socket, upgradeHead) => {
+				if (parseUrl(request.url).pathname === `/${restEndpoint}/push`) {
+					wsServer.handleUpgrade(request, socket, upgradeHead, (ws) => {
+						request.ws = ws;
+
+						const response = new ServerResponse(request);
+						response.writeHead = (statusCode) => {
+							if (statusCode > 200) ws.close();
+							return response;
+						};
+
+						// @ts-expect-error `handle` isn't documented
+						// eslint-disable-next-line @typescript-eslint/no-unsafe-call
+						app.handle(request, response);
+					});
+				}
+			});
+		}
+	}
+
+	/** Sets up the push endppoint that the frontend connects to. */
+	setupPushHandler(restEndpoint: string, app: Application) {
+		app.use(
+			`/${restEndpoint}/push`,
+			// eslint-disable-next-line @typescript-eslint/unbound-method
+			this.authService.authMiddleware,
+			(req: SSEPushRequest | WebSocketPushRequest, res: PushResponse) =>
+				this.handleRequest(req, res),
+		);
+	}
+
 	handleRequest(req: SSEPushRequest | WebSocketPushRequest, res: PushResponse) {
 		const {
 			ws,
 			query: { pushRef },
 			user,
+			headers,
 		} = req;
 
+		let connectionError = '';
 		if (!pushRef) {
+			connectionError = 'The query parameter "pushRef" is missing!';
+		} else if (
+			inProduction &&
+			headers.origin &&
+			headers.host &&
+			!headers.origin.endsWith(headers.host)
```

Comment:
Is checking `endsWith` enough here? Could we have a situation where the origin is something like `eviln8n.io`, matching to the host's `n8n.io`?

Should we also consider port?

Reply:
the port is included in both these headers 
![image](https://github.com/user-attachments/assets/089a89b9-00b7-4993-9db7-e22fe1152ffc).
So, `endsWith` is fine. The only way this could be better would be to check that origin is `http://host` or `https://host`